### ISSUE TITLE
Error in data$construct() :    lgb.Dataset.construct: cannot create Dataset handle

### ISSUE THREAD

**Issue by dwh2102** [Microsoft/LightGBM#1730]
Tried solutions described here: https://github.com/Microsoft/LightGBM/issues/571

Did not work. Re-installed lightgbm, re-started R, used lgb.unloader and rec'd same error.  Any help is appreciated--thanks!

## Environment info

Operating System: Windows 7 Professional

CPU/GPU model: Intel(R) Core(TM) i3-4150 CPU @ 3.50 GHz 3.50 GHz

C++/Python/R version: RStudio Version 1.1.456

## Error message

Error in data$construct() : 
  lgb.Dataset.construct: cannot create Dataset handle

## Reproducible examples

require(lightgbm)
> require(methods)
> data(agaricus.train, package = "lightgbm")
> data(agaricus.test, package = "lightgbm")
> train <- agaricus.train
> test <- agaricus.test
> print("Training lightgbm with sparseMatrix")
[1] "Training lightgbm with sparseMatrix"
> bst <- lightgbm(data = train$data,
+                 label = train$label,
+                 num_leaves = 4,
+                 learning_rate = 1,
+                 nrounds = 2,
+                 objective = "binary")
Error in data$construct() : 
  lgb.Dataset.construct: cannot create Dataset handle
> lgb.unloader(wipe = TRUE)
> library(lightgbm)
> bst <- lightgbm(data = train$data,
+                 label = train$label,
+                 num_leaves = 4,
+                 learning_rate = 1,
+                 nrounds = 2,
+                 objective = "binary")
Error in data$construct() : 
  lgb.Dataset.construct: cannot create Dataset handle

## Steps to reproduce

1.
2.
3.

**Comment by StrikerRUS** [comment]
ping @jameslamb and @Laurae2

**Comment by jameslamb** [comment]
Thank you for the reproducible example! I will see if I can reproduce it on my machine and look for a fix

**Comment by dwh2102** [comment]
Great! Thanks for your help.

Dave

**Comment by jameslamb** [comment]
Looking into this today. Will update shortly

**Comment by jameslamb** [comment]
Ok I ran the code from above on my Mac and could not reproduce the issue.

```
library(lightgbm)
library(methods)

data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")

train <- agaricus.train
test <- agaricus.test

print("Training lightgbm with sparseMatrix")

bst <- lightgbm(
    data = train$data,
    label = train$label,
    num_leaves = 4,
    learning_rate = 1,
    nrounds = 2,
    objective = "binary"
)
```

I got this result:

```
[LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 3140, number of negative: 3373
[LightGBM] [Info] Total Bins 128
[LightGBM] [Info] Number of data: 6513, number of used features: 107
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.482113 -> initscore=-0.071580
[LightGBM] [Info] Start training from score -0.071580
[1]:	train's binary_logloss:0.198597 
[2]:	train's binary_logloss:0.111535 
```

**Comment by dwh2102** [comment]
Thanks for looking into this for me--any idea of what my issue might be or
what I should try next? Thanks!

**Comment by jameslamb** [comment]
@dwh2102 Now that I've got it working on my Mac, the next thing I'll try (probably tonight) is to build on a Windows machine.

I was looking through the `construct()` code last night and there are about different places this could have failed. If nothing else, I'll add logging in there so we can see which block executes for you. Then if that doesn't help us find an answer we can pull in @Laurae2 to help.

We won't abandon you! Thanks for your patience.

**Comment by dwh2102** [comment]
Great! Thank you so much.

**Comment by Laurae2** [comment]
@jameslamb Windows 10, R 3.5.0, LightGBM master:

```r
R version 3.5.0 (2018-04-23) -- "Joy in Playing"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(lightgbm)
Loading required package: R6
> library(methods)
> 
> data(agaricus.train, package = "lightgbm")
> data(agaricus.test, package = "lightgbm")
> 
> train <- agaricus.train
> test <- agaricus.test
> 
> print("Training lightgbm with sparseMatrix")
[1] "Training lightgbm with sparseMatrix"
> 
> bst <- lightgbm(
+     data = train$data,
+     label = train$label,
+     num_leaves = 4,
+     learning_rate = 1,
+     nrounds = 2,
+     objective = "binary"
+ )
Loading required package: Matrix
[LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 3140, number of negative: 3373
[LightGBM] [Info] Total Bins 128
[LightGBM] [Info] Number of data: 6513, number of used features: 107
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.482113 -> initscore=-0.071580
[LightGBM] [Info] Start training from score -0.071580
[1]:	train's binary_logloss:0.198597 
[2]:	train's binary_logloss:0.111535 
> sessionInfo()
R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Matrix_1.2-14  lightgbm_2.2.1 R6_2.2.2      

loaded via a namespace (and not attached):
[1] compiler_3.5.0    magrittr_1.5      tools_3.5.0       yaml_2.1.19       grid_3.5.0        data.table_1.11.6 jsonlite_1.5      lattice_0.20-35  
```

**Comment by jameslamb** [comment]
The plot thickens!

@Laurae2 I'm going to make the PR to improve logging anyway, then maybe if @dwh2102 can run and reproduce the problem, we'll be able to form a better hypothesis about where something went wrong.

**Comment by aaliwei91** [comment]
Hi, I have the same problem. 

System: windows 7 

library: I compile the source code with VS 2017 to generate a .dll file and then compile to a zip file

Then I install the package through the zip file from RStudio, and run the code
`> library(lightgbm, quietly=TRUE)
> data(agaricus.train, package = "lightgbm")
> train <- agaricus.train
> dtrain <- lgb.Dataset(train$data, label = train$label)
Loading required package: Matrix
> lgb.Dataset.construct(dtrain)
Error in dataset$construct() : 
  lgb.Dataset.construct: cannot create Dataset handle
`
Maybe there are some problems during the compiling and installation of the package. I remember after I run the R CMD INSTALL command, it shows "DONE", however  I cannot find any function in the lightgbm library (I use lightgbm:: and there are only 3 datasets without functions). Then I make the directory into zip file and then install the zip file, and then find the functions at last. However, when I run the code above it shows the problem.

Could you use my .zip file and check what the problem is? Thank you,

[lightgbm_2.2.2.zip](https://github.com/Microsoft/LightGBM/files/2509334/lightgbm_2.2.2.zip)

**Comment by aaliwei91** [comment]
> 
> 
> Hi, I have the same problem.
> 
> System: windows 7
> 
> library: I compile the source code with VS 2017 to generate a .dll file and then compile to a zip file
> 
> Then I install the package through the zip file from RStudio with Windows R Open 3.5.1
> 
> platform x86_64-w64-mingw32
> arch x86_64
> os mingw32
> system x86_64, mingw32
> status
> major 3
> minor 5.1
> year 2018
> month 07
> day 02
> svn rev 74947
> language R
> version.string R version 3.5.1 (2018-07-02)
> nickname Feather Spray
> , and run the code
> 
> ```
> > library(lightgbm, quietly=TRUE)
> > data(agaricus.train, package = "lightgbm")
> > train <- agaricus.train
> > dtrain <- lgb.Dataset(train$data, label = train$label)
> Loading required package: Matrix
> > lgb.Dataset.construct(dtrain)
> Error in dataset$construct() : 
>   lgb.Dataset.construct: cannot create Dataset handle
> ```
> 
> Maybe there are some problems during the compiling and installation of the package. I remember after I run the R CMD INSTALL command, it shows "DONE", however I cannot find any function in the lightgbm library (I use lightgbm:: and there are only 3 datasets without functions). Then I make the directory into zip file and then install the zip file, and then find the functions at last. However, when I run the code above it shows the problem.
> 
> For another case, I use code from
> 
> https://www.kaggle.com/nschneider/gbm-vs-xgboost-vs-lightgbm
> 
> it shows:
> 
> ```
> > lgb.bench = microbenchmark(
> +   lgb.model <- lgb.train(
> +     params = params.lgb
> +     , data = lgb.train
> +     , valids = list(test = lgb.test)
> +     , learning_rate = 0.1
> +     , num_leaves = 7
> +     , num_threads = 2
> +     , nrounds = 500
> +     , early_stopping_rounds = 40
> +     , eval_freq = 20
> +   )
> +   , times = 5L
> + )
> [LightGBM] [Fatal] Check failed: num_data > 0 at d:\lightgbm\src\io\dataset.cpp, line 27 .
> 
> Error in data$construct() : 
>   lgb.Dataset.construct: cannot create Dataset handle
> ```
> 
> Could you use my .zip file and check what the problem is? Thank you,
> 
> [lightgbm_2.2.2.zip](https://github.com/Microsoft/LightGBM/files/2509334/lightgbm_2.2.2.zip)

Now I solve this problem by compiling the source code again to generate the .dll with command
```
cmake -DCMAKE_GENERATOR_PLATFORM=X64 -DUSE_R35=1 ..
cmake --build . --target _lightgbm --config Release

```
Then I copied the .dll files to the libs directory of R-package

and add the directory to a zip file, then install it from RStudio
I've uploaded the workable lightgbb.zip and you can use it directly for Windows.
[lightgbm.zip](https://github.com/Microsoft/LightGBM/files/2512975/lightgbm.zip)

**Comment by StrikerRUS** [comment]
@jameslamb @Laurae2 any updates here?

**Comment by Laurae2** [comment]
@aaliwei91 Using your lightgbm.zip file on my side, it works.

Did you make sure to not have LightGBM loaded when installing LightGBM? Make sure to:

* Compile/Write the package to the right .libPaths() folder
* Install LightGBM from a fresh R session
* Try using Rgui as you are using Windows. RStudio has issues sometimes with installing packages with specific dependencies.

Log using your lightgbm.zip:

```r
> library(lightgbm)
Loading required package: R6
Warning message:
package ‘lightgbm’ was built under R version 3.5.1 
> data(agaricus.train, package = "lightgbm")
> train <- agaricus.train
> dtrain <- lgb.Dataset(train$data, label = train$label)
Loading required package: Matrix
> data(agaricus.test, package = "lightgbm")
> test <- agaricus.test
> dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
> params <- list(objective = "regression", metric = "l2")
> valids <- list(test = dtest)
> model <- lgb.train(params,
+                    dtrain,
+                    100,
+                    valids,
+                    min_data = 1,
+                    learning_rate = 1,
+                    early_stopping_rounds = 10)
[LightGBM] [Info] Total Bins 137
[LightGBM] [Info] Number of data: 6513, number of used features: 116
[LightGBM] [Info] Start training from score 0.482113
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[2]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[3]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[4]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[5]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[6]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[7]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[8]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[9]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[10]:	test's l2:6.44165e-17 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[11]:	test's l2:6.44165e-17 
```

**Comment by Laurae2** [comment]
@aaliwei91 any news?

**Comment by StrikerRUS** [comment]
@dwh2102 Are you still encountering this issue?

**Comment by dwh2102** [comment]
No longer an issue for me. Thanks!

**Comment by Josie-xufan** [comment]
Hi, I have some questions about the lightGBM package.

system: Windows10

I downloaded the lightGBM package from the internet. And then I run the code to use the package. But there is one error. I don't know how to solve it. Can you help me have a look at the code and check what is the problem. Thanks a lot!

Here is the R-package I used.
[R-package.zip](https://github.com/microsoft/LightGBM/files/4052659/R-package.zip)

And this is my code.
install.packages("E:\\LightGBM\\R-package", repos = NULL, type="source",lib = "E:\\R3.6.2\\library",dependencies = TRUE) 
library(lightgbm)

bia1 <- data.matrix(traindata[,-22]) # change class(Vars) into numeric
bia2 <- traindata$REAP.CD272.ACCGAACA
dtrain <- lgb.Dataset(data=bia1,label=bia2)

bia3 <- data.matrix(testdata[,-22]) # change class(Vars) into numeric
bia4 <- testdata$REAP.CD272.ACCGAACA
dtest <- lgb.Dataset.create.valid(dataset=dtrain,data=bia3,label=bia4)

valids <- list(test=dtest)
params <- list(objective = "regression", metric = "l2") # L2 not twelve

lgb1 <- lgb.train(params=params,
                  data=dtrain,
                  valids = valids,
                  min_data =1, # min data in a group
                  learning_rate=0.1, # smaller,slower,maybe more accurate
                  nrounds = 300,
                  early_stopping_rounds = 20) #if not better than last 20 rounds,stop

But the result is this.
Error in reference$get_colnames() : 
  dim: cannot get dimensions before dataset has been constructed, please call lgb.Dataset.construct explicitly

**Comment by jameslamb** [comment]
@Josie-xufan can you please open a new issue? (https://github.com/microsoft/LightGBM/issues)

Your issue looks different from the others in this thread, and I think it would be best to talk in a new issue. We also need some other information which you'll see requested in our issue template, including your `lightgbm` version.

Once you open the issue I'll be happy to help out!
